dds/idl/marshal_generator.cpp

```diff
@@ -769,567 +1108,253 @@ namespace {
-    string const_cxx = cxx;
-    if (use_cxx11) {
-      be_global->header_ << "struct " << sf.underscored_ << "_tag {};\n\n";
-      const_cxx = "IDL::DistinctType<const " + cxx + ", " + sf.underscored_ + "_tag>";
-    } else {
-      const_cxx = "const " + cxx + '&';
+
+    AST_Type* elem = resolveActualType(arr->base_type());
+    TryConstructFailAction try_construct = be_global->array_element_try_construct(arr);
+    Classification elem_cls = classify(elem);
+    const bool primitive = elem_cls & CL_PRIMITIVE;
+    if (!elem->in_main_file()
+        && elem->node_type() != AST_Decl::NT_pre_defined) {
+      be_global->add_referenced(elem->file_name().c_str());
+    }
+    const std::string cxx_elem =
+      anonymous ? anonymous->scoped_elem_ : scoped(arr->base_type()->name());
+    size_t n_elems = 1;
+    for (size_t i = 0; i < arr->n_dims(); ++i) {
+      n_elems *= arr->dims()[i]->ev()->u.ulval;
     }
+
+    base_wrapper.generate_tag();
+
+    if (!nested_key_only) {
+      Wrapper wrapper(base_wrapper);
+      wrapper.is_const_ = false;
+      wrapper.done();
+      Function set_default("set_default", "void");
+      set_default.addArg("arr", wrapper.wrapped_type_name());
+      set_default.endArgs();
+      string indent = "  ";
+      NestedForLoops nfl("CORBA::ULong", "i", arr, indent);
+      be_global->impl_ <<
+        indent << type_to_default(elem, wrapper.value_access() + nfl.index_);
+    }
+
     {
+      Wrapper wrapper(base_wrapper);
+      wrapper.done();
       Function serialized_size("serialized_size", "void");
       serialized_size.addArg("encoding", "const Encoding&");
       serialized_size.addArg("size", "size_t&");
-      serialized_size.addArg(sf.arg_.c_str(), sf.const_ref_);
+      serialized_size.addArg("arr", wrapper.wrapped_type_name());
       serialized_size.endArgs();
 
       generate_dheader_code("    serialized_size_delimiter(encoding, size);\n", !primitive, false);
 
-      be_global->impl_ << sf.const_unwrap_ <<
-        "  OpenDDS::DCPS::primitive_serialized_size_ulong(encoding, size);\n"
-        "  if (" << check_empty << ") {\n"
-        "    return;\n"
-        "  }\n";
-      if (sf.as_cls_ & CL_ENUM) {
-        be_global->impl_ <<
-          "  OpenDDS::DCPS::primitive_serialized_size_ulong(encoding, size, " + get_length + ");\n";
-      } else if (sf.as_cls_ & CL_PRIMITIVE) {
-        be_global->impl_ << checkAlignment(sf.as_act_) <<
-          "  " + getSizeExprPrimitive(sf.as_act_, get_length) << ";\n";
-      } else if (sf.as_cls_ & CL_INTERFACE) {
+      if (elem_cls & CL_ENUM) {
         be_global->impl_ <<
-          "  // sequence of objrefs is not marshaled\n";
-      } else if (sf.as_cls_ == CL_UNKNOWN) {
+          "  OpenDDS::DCPS::primitive_serialized_size_ulong(encoding, size);\n";
+        if (n_elems > 1) {
+          be_global->impl_ <<
+            "  OpenDDS::DCPS::primitive_serialized_size_ulong(encoding, size, "
+              << n_elems - 1 << ");\n";
+        }
```

Comment:
Why not just generate a single call with 3 arguments (and the third may be a `1`)?